finality-aleph/src/justification/compatibility.rs

```diff
@@ -27,22 +32,242 @@ impl From<AlephJustificationV1> for AlephJustification {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub enum JustificationDecoding {
+enum VersionedAlephJustification {
+    // Most likely from the future.
+    Other(u32, Vec<u8>),
     V1(AlephJustificationV1),
     V2(AlephJustification),
-    Err,
 }
 
-pub fn backwards_compatible_decode(justification_raw: Vec<u8>) -> JustificationDecoding {
+impl Encode for VersionedAlephJustification {
+    fn size_hint(&self) -> usize {
+        use VersionedAlephJustification::*;
+        let version_and_bytes_size = 2 * size_of::<u32>();
+        version_and_bytes_size
+            + match self {
+                Other(_, payload) => payload.size_hint(),
+                V1(justification) => justification.size_hint(),
+                V2(justification) => justification.size_hint(),
+            }
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        use VersionedAlephJustification::*;
+        match self {
+            Other(version, payload) => {
+                let mut result = version.encode();
+                let mut payload = payload.encode();
+                let num_bytes = payload.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut payload);
+                result
+            }
+            V1(justification) => {
+                let mut result = (1_u32).encode();
+                let mut justification = justification.encode();
+                let num_bytes = justification.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut justification);
+                result
+            }
+            V2(justification) => {
+                let mut result = (2_u32).encode();
+                let mut justification = justification.encode();
+                let num_bytes = justification.len() as u32;
+                result.append(&mut num_bytes.encode());
+                result.append(&mut justification);
+                result
+            }
+        }
+    }
+}
+
+impl Decode for VersionedAlephJustification {
+    fn decode<I: CodecInput>(input: &mut I) -> Result<Self, CodecError> {
+        use VersionedAlephJustification::*;
+        let version = u32::decode(input)?;
+        let num_bytes = u32::decode(input)?;
+        match version {
+            1 => Ok(V1(AlephJustificationV1::decode(input)?)),
+            2 => Ok(V2(AlephJustification::decode(input)?)),
+            _ => {
+                // TODO: figure out some limit for `num_bytes` and error out if this alloc would be
+                // too big
```

Comment:
This does not seem like a TODO we would like to leave for later :P 

One signature is `64` bytes, so if we stay below `100` nodes, then something like `2^{13}` should be safe. 
Probably before we scale beyond `100` we will move to succinct multisigs, so the const will never need to grow, I guess?

Reply:
Hey, cool, that would mean I can just switch this to a `u16` and that would be acceptable, wouldn't it?

Reply:
Yes, and I believe the same can be done for version :)

Reply:
Done, also rebased on new main.